binaries/cli/src/command/topic/selector.rs

```diff
@@ -0,0 +1,161 @@
+use std::{borrow::Cow, collections::HashMap, fmt, net::IpAddr};
+
+use crate::{
+    LOCALHOST,
+    common::{connect_to_coordinator, resolve_dataflow_identifier},
+};
+use communication_layer_request_reply::TcpRequestReplyConnection;
+use dora_core::{
+    config::InputMapping, descriptor::Descriptor, topics::DORA_COORDINATOR_PORT_CONTROL_DEFAULT,
+};
+use dora_message::{
+    DataflowId,
+    cli_to_coordinator::ControlRequest,
+    coordinator_to_cli::ControlRequestReply,
+    id::{DataId, NodeId},
+};
+use eyre::{Context, ContextCompat, bail};
+use uuid::Uuid;
+
+#[derive(Debug, clap::Args)]
+pub struct DataflowSelector {
+    /// Identifier of the dataflow
+    #[clap(long, short, value_name = "UUID_OR_NAME")]
+    pub dataflow: Option<String>,
+    /// Address of the dora coordinator
+    #[clap(long, value_name = "IP", default_value_t = LOCALHOST)]
+    pub coordinator_addr: IpAddr,
+    /// Port number of the coordinator control server
+    #[clap(long, value_name = "PORT", default_value_t = DORA_COORDINATOR_PORT_CONTROL_DEFAULT)]
+    pub coordinator_port: u16,
+}
+
+impl DataflowSelector {
+    pub fn resolve(&self) -> eyre::Result<(Box<TcpRequestReplyConnection>, Uuid, Descriptor)> {
+        let mut session =
+            connect_to_coordinator((self.coordinator_addr, self.coordinator_port).into())
+                .wrap_err("failed to connect to dora coordinator")?;
+        let dataflow_id = resolve_dataflow_identifier(&mut *session, self.dataflow.as_deref())?;
+        let reply_raw = session
+            .request(
+                &serde_json::to_vec(&ControlRequest::Info {
+                    dataflow_uuid: dataflow_id,
+                })
+                .unwrap(),
+            )
+            .wrap_err("failed to send message")?;
+        let reply: ControlRequestReply =
+            serde_json::from_slice(&reply_raw).wrap_err("failed to parse reply")?;
+        match reply {
+            ControlRequestReply::DataflowInfo { descriptor, .. } => {
+                Ok((session, dataflow_id, descriptor))
+            }
+            ControlRequestReply::Error(err) => bail!("{err}"),
+            other => bail!("unexpected list dataflow reply: {other:?}"),
+        }
+    }
+}
+
+#[derive(Debug, clap::Args)]
+pub struct TopicSelector {
+    #[clap(flatten)]
+    pub dataflow: DataflowSelector,
+    /// Data to inspect, e.g. `node_id/output_id`
+    #[clap(value_name = "DATA")]
+    pub data: Vec<String>,
+}
+
+#[derive(Clone, PartialOrd, Ord, PartialEq, Eq)]
+pub struct TopicIdentifier {
+    pub node_id: NodeId,
+    pub data_id: DataId,
+}
+
+impl fmt::Display for TopicIdentifier {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}/{}", self.node_id, self.data_id)
+    }
+}
+
+impl TopicSelector {
+    pub fn resolve(
+        &self,
+    ) -> eyre::Result<(
+        Box<TcpRequestReplyConnection>,
+        (DataflowId, Vec<TopicIdentifier>),
```

Comment:
This function should take a `&mut TcpRequestReplyConnection` argument instead of initializing a session and returning it. This way, the return type becomes simpler and the function can be called multiple times.